### ISSUE TITLE
Cant register with scheme set to DIGEST

### ISSUE THREAD

**Issue by gwozniak** [LubosD/twinkle#36]
Hi,
I found another issue in Twinkle, when proxy responses 401 with Authentication Scheme set to 
DIGEST (capitals) Twinkle do nothing and  only writes to log:
++ 3-12-2015 18:36:13.718998 INFO NORMAL t_auth::authorize
Unsupported authentication scheme: DIGEST

When I changed auth.cpp
```
-       if (c.auth_scheme != AUTH_DIGEST) {
+       if (cmp_nocase(chlg.auth_scheme, AUTH_DIGEST) != 0) {
```
Twinkle started popup an login windows but with realm  set  to ""  while proxy responses with realm set to domain name.
```
REGISTER sip:windmobile.pl SIP/2.0
Via: SIP/2.0/UDP 192.168.198.1;rport;branch=z9hG4bKyjeiqtva
Max-Forwards: 70
To: <sip:914165001@windmobile.pl>
From: <sip:914165001@windmobile.pl>;tag=jdilu
Call-ID: vshuiexrpoapovk@skin.skin.local
CSeq: 136 REGISTER
Contact: <sip:914165001@192.168.198.1>;expires=3600
Allow: INVITE,ACK,BYE,CANCEL,OPTIONS,PRACK,REFER,NOTIFY,SUBSCRIBE,INFO,MESSAGE
User-Agent: Twinkle/1.9.0
Content-Length: 0

SIP/2.0 401 Unauthorized
Via:SIP/2.0/UDP 192.168.198.1;branch=z9hG4bKyjeiqtva;rport
From:<sip:914165001@windmobile.pl>;tag=jdilu
To:<sip:914165001@windmobile.pl>;tag=1177029366-1449167798189
Call-ID:vshuiexrpoapovk@skin.skin.local
CSeq:136 REGISTER
WWW-Authenticate:DIGEST qop="auth",nonce="BroadWorksXihql39gtTi2cq07BW",realm="windmobile.pl",algorithm=MD5
Content-Length:0

REGISTER sip:windmobile.pl SIP/2.0
Via: SIP/2.0/UDP 192.168.198.1;rport;branch=z9hG4bKqttxolng
Max-Forwards: 70
To: <sip:914165001@windmobile.pl>
From: <sip:914165001@windmobile.pl>;tag=jdilu
Call-ID: vshuiexrpoapovk@skin.skin.local
CSeq: 137 REGISTER
Contact: <sip:914165001@192.168.198.1>;expires=3600
Authorization: Digest username="914165001",uri="sip:windmobile.pl",response="b072626c72888ff301e7aa420eca7759",algorithm=MD5
Allow: INVITE,ACK,BYE,CANCEL,OPTIONS,PRACK,REFER,NOTIFY,SUBSCRIBE,INFO,MESSAGE
User-Agent: Twinkle/1.9.0
Content-Length: 0

SIP/2.0 401 Unauthorized
Via:SIP/2.0/UDP 192.168.198.1;branch=z9hG4bKqttxolng;rport
From:<sip:914165001@windmobile.pl>;tag=jdilu
To:<sip:914165001@windmobile.pl>;tag=1031875026-1449167839419
Call-ID:vshuiexrpoapovk@skin.skin.local
CSeq:137 REGISTER
WWW-Authenticate:DIGEST qop="auth",nonce="BroadWorksXihql45a3Tahuuj5BW",realm="windmobile.pl",algorithm=MD5
Content-Length:0


```

**Comment by gwozniak** [comment]
Hi, I manged to rapair this digest issue:
```
diff --git a/src/auth.cpp b/src/auth.cpp
index ff5c4bc..4cc7a4a 100644
--- a/src/auth.cpp
+++ b/src/auth.cpp
@@ -134,7 +134,8 @@ bool t_auth::authorize(t_user *user_config, t_request *r, t_response *resp) {
        }
 
        // Only DIGEST is supported
-       if (c.auth_scheme != AUTH_DIGEST) {
+       //if (c.auth_scheme != AUTH_DIGEST) {
+       if (cmp_nocase(c.auth_scheme, AUTH_DIGEST) != 0) {
                log_file->write_header("t_auth::authorize");
                log_file->write_raw("Unsupported authentication scheme: ");
                log_file->write_raw(c.auth_scheme);
diff --git a/src/parser/challenge.cpp b/src/parser/challenge.cpp
index 9c365a5..a15c696 100644
--- a/src/parser/challenge.cpp
+++ b/src/parser/challenge.cpp
@@ -163,7 +163,8 @@ string t_challenge::encode(void) const {
        string s = auth_scheme;
        s += ' ';
 
-       if (auth_scheme == AUTH_DIGEST) {
+       //if (auth_scheme == AUTH_DIGEST) {
+       if (cmp_nocase(auth_scheme,AUTH_DIGEST)==0) {
                s += digest_challenge.encode();
        } else {
                for (list<t_parameter>::const_iterator i = auth_params.begin();
diff --git a/src/parser/credentials.cpp b/src/parser/credentials.cpp
index 067744c..0ada6af 100644
--- a/src/parser/credentials.cpp
+++ b/src/parser/credentials.cpp
@@ -143,7 +143,8 @@ string t_credentials::encode(void) const {
        string s = auth_scheme;
        s += ' ';
 
-       if (auth_scheme == AUTH_DIGEST) {
+       //if (auth_scheme == AUTH_DIGEST) {
+       if (cmp_nocase(auth_scheme,AUTH_DIGEST)==0) {
                s += digest_response.encode();
        } else {
                for (list<t_parameter>::const_iterator i = auth_params.begin();
diff --git a/src/parser/scanner.lxx b/src/parser/scanner.lxx
index 19d9f11..b86ba83 100644
--- a/src/parser/scanner.lxx
+++ b/src/parser/scanner.lxx
@@ -302,7 +302,7 @@ WORD_SYM    [[:alnum:]\-\.!%\*_\+\`\'~\(\)<>:\\\"\/\[\]\?\{\}]
 <C_NEW>\n              { return T_CRLF; }
 
        /* Authorization scheme */
-<C_AUTH_SCHEME>Digest          { return T_AUTH_DIGEST; }
+<C_AUTH_SCHEME>[Dd][Ii][Gg][Ee][Ss][Tt]        { return T_AUTH_DIGEST; }
 <C_AUTH_SCHEME>{TOKEN_SYM}+    { yylval.yyt_str = new string(yytext);
                                  MEMMAN_NEW(yylval.yyt_str);
                                  return T_AUTH_OTHER; }

```

**Comment by mkubecek** [comment]
I believe this is correct as both RFC 2617 and RFC 7235 state that the authentication scheme is to be matched case insensitively. If you are going to do a pull request, please consider some cleanup:
* commenting out old version of a line doesn't make sense in the git age, it just clutters the code
* please add spaces around "`==`" operator
* flex seems to support the "`(?i:digest)`" syntax for case insensitive matching (but I'm not sure how portable it is)

**Comment by gwozniak** [comment]
fix merged in #55, thanks :)
